File: Resource/Graph/타겟넘버.py
def solution(numbers, target):
    answer = 0

    # dfs 현재 만들어진 수랑 더해진 수(len) 매개변수로 넘기기
    def dfs(number, index):
        nonlocal answer

        # 종료 조건 len == len(numbers)
        # 종료 조건에서 sum == target answer += 1
        if index == len(numbers):
            if number == target:
                answer += 1
            return

        # dfs +, - 넣기
        dfs(number + numbers[index], index + 1, )
        dfs(number - numbers[index], index + 1, )

    dfs(0, 0)

    return answer


print(solution([4, 1, 2, 1], 4))
print(solution([1, 1, 1, 1, 1], 3))

File: cpp/타겟넘버.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

void dfs(const vector<int>& numbers, int target, int index, int sum, int& answer){
    // 1. 탈출 조건(마지막 숫자까지 모두 확인한 후)
    if (index == numbers.size()) {
        if (sum == target) {
          answer++;
        }
        return;
    }
    
    // 2. 현재 숫자를 더하고 빼고 두 가지 경우로 dfs 재귀호출
    dfs(numbers, target, index + 1, sum + numbers[index], answer);
    dfs(numbers, target, index + 1, sum - numbers[index], answer);  
}

int solution(vector<int> numbers, int target) {
    int answer = 0;
    dfs(numbers, target, 0, 0, answer);
    return answer;
}

// 로컬에서 테스트를 위한 main 함수
int main() {
    vector<int> numbers = {1, 1, 1, 1, 1};
    int target = 3;
    int result = solution(numbers, target);
    
    // 결과 출력
    cout << "결과: " << result << endl; // 예상 결과: 5
  
    return 0;
}
